[BE] feat: 관리자 히어릿 추가 시 S3 연동 #208

yuyoungrhee opened the pull request:
<!-- PR 제목 예시
ex. [BE] 로그인 API 구현
ex. [AN] 페이지 구현 -->

## 📌 변경 내용 & 이유
<!-- ex. 
- 사용자 로그인 기능 구현
- 로그인 실패 시 에러 메시지 반환 처리 추가 -->
- 관리자페이지에서 히어릿 추가 시 S3 연동하여 S3업로드 + DB 반영 동시에 되도록 구현
- 파일을 직접 업로드하는 것으로 변경
- 검증 로직 추가: 파일 prefix(ORG, SHR, SCR)이 맞게 업로드 되어있는지 / Keyword 제외 모든 값이 존재하는지 검증

## 📸 스크린샷 (선택)
> UI 변경이 있을 경우 첨부
<img width="1487" height="776" alt="image" src="https://github.com/user-attachments/assets/e3cfeaa5-b1a9-4ed9-8355-32e3de2e78c6" />

## 🧪 테스트 방법 (선택)
<!-- 코드 리뷰하는 팀원이 돌려볼 수 있도록 작성해주기 
ex. ReservationTest의 XXXX 테스트메서드로 확인 
ex. 어떤화면의 어떤기능에서 확인 -->
테스트 못 함... S3의 저주......기도합시다.

.env 파일 바꿔야합니다 노션에 업로드해둘게요

## 📢 논의하고 싶은 내용
<!-- 이러이러해서 이부분을 이러이러하게 구현했는데 이방식이 괜찮은지? -->

## 🧩 관련 이슈
<!-- 이슈 태그: #123 -->
<!-- 이슈 닫기: close #123 -->
close #208


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **신규 기능**
  * 관리자 페이지에 히어릿(오디오, 스크립트 등) 업로드 기능이 추가되었습니다. 파일 업로드 시 카테고리 및 키워드 선택이 가능합니다.
  * AWS S3 연동을 통한 파일 저장 기능이 도입되었습니다.
  * 업로드 파일명 및 형식에 대한 유효성 검사가 적용되었습니다.
  * 히어릿 상세 조회 시 키워드 ID와 이름이 함께 제공됩니다.

* **버그 수정**
  * 카테고리, 키워드, 회원가입, 북마크 생성 시 응답 헤더에 Location이 포함되지 않도록 수정되었습니다.

* **UI/UX 개선**
  * 히어릿 추가 폼이 별도의 페이지로 분리되고, 네비게이션 바에 "히어릿 업로드" 메뉴가 추가되었습니다.
  * 기존 히어릿 목록 페이지에서 업로드 폼이 제거되었습니다.
  * 템플릿의 네비게이션 바 포함 방식이 개선되었습니다.
  * 관리자 홈 페이지 타이틀 및 메뉴 문구가 일부 수정되었습니다.

* **테스트**
  * 파일 업로드 방식을 반영한 테스트 코드 및 테스트 리소스가 추가되었습니다.

* **설정**
  * S3 연동, 파일 업로드 용량 제한 등 환경설정이 추가 및 개선되었습니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

yuyoungrhee committed:
[BE] feat: 관리자 히어릿 추가 시 파일 첨부하여 S3에 동시 업로드 #208

yuyoungrhee committed:
[BE] Merge branch 'origin/develop-be' into feat/#208-admin-s3-integration #208

yuyoungrhee committed:
[BE] test: 관리자 히어릿 업로드 방식 변경에 따른 테스트 수정 #208

yuyoungrhee committed:
[BE] fix: S3 업로드 경로 수정 #208

yuyoungrhee committed:
[BE] refactor: S3 관련 Exception 추가 #208

Byesol commented on backend/src/main/java/com/onair/hearit/admin/application/S3Uploader.java:
```diff
@@ -0,0 +1,52 @@
+    private String bucket;
+
+    public String uploadOriginalAudio(MultipartFile multipartFile) {
+        String key = "hearit/audio/" + multipartFile.getOriginalFilename();
```
이런 url은 상수로 빼는 게 좋아보여요

yuyoungrhee replied:
재사용성 / 의미명시/유지보수성 관점에서 생각해봤을 때, 일회성 상수는 빼지 않는 것이 더 좋은 것 같습니다. 어떻게 생각하시나요? 더군다나 한곳에서 사용되는 상수라면 저는 보통 분리하지 않았습니다. 

재사용성 - 재사용 안 되고 uploadOriginalAudio 안에서만 의미있음
의미 명시 - 이거를 대문자 상수명을 따로 짓는 것보다 현재가 더 가독성이 좋은 거 같음(이거는 개인적)
유지보수성 - 상수로 빼나 안 빼나 똑같음 (결국 한 군데만 고치면 되니까)

Byesol replied:
일단 해당 "hearit/audio/" 은  아래의 `uploadShortAudio` 에서도 똑같이 쓰이고 있긴 해요. 따라서 저는 개인적으로 'AUDIO_DIR_PREFIX' 로 두 군데서 사용한다면 저는 개인적으로 좀 더 의미 파악도 쉽고, 두 군데서 사용하니 
유지보수성 측면에서도 좋을 것 같느데 어떻게 생각하시나요?

yuyoungrhee replied:
아 동일하게 쓰이는 데가 있었네요 그러면 인정합니다

가콩의 피드백을 반영해서 enum으로 관리해서 간단하게 변경해봤어용
[[BE] refactor: 파일 업로드 및 검증 로직 리팩터링](https://github.com/woowacourse-teams/2025-hEARit/pull/216/commits/190a39f5761454550af4fbb0e2fe48423cd0d31f) https://github.com/woowacourse-teams/2025-hEARit/issues/208

Byesol commented on backend/src/test/resources/test_script.json:
```diff
@@ -0,0 +1,14 @@
+[
```
이 테스트 파일들은 이제 제거해도 되지 않나요?

yuyoungrhee replied:
이 테스트 파일이 있어야 통합테스트가 돌아갑니다..! 아니면 mocking 할수밖에없어요

`AdminCategoryControllerTest`의 `createHearit()` 테스트요

Byesol replied:
아하. 이해했습니다! 감사합니다

Byesol reviewed (commented):
고생 많으셨어요!

yuyoungrhee committed:
[BE] refactor: navbar 삽입 문법 최신화 #208

- 타임리프 구버전 문법을 쓰고있어 WARN 로깅이 되어 수정함

yuyoungrhee committed:
[BE] refactor: 히어릿 업로드시 검증 로직 추가 #208

- 파일 이름 검증 및 값 존재 검증

yuyoungrhee committed:
[BE] Merge branch 'develop-be' into feat/#208-admin-s3-integration #208

yuyoungrhee commented on backend/src/main/java/com/onair/hearit/common/exception/GlobalExceptionHandler.java:
```diff
@@ -38,7 +38,7 @@ public ResponseEntity<Object> handleNoResourceFoundException(NoResourceFoundExce
     @ExceptionHandler(Exception.class)
     public ProblemDetail handleUnhandledException(Exception ex, HttpServletRequest request) {
-        log.error("서버 내부 오류", ex);
+        log.error("서버 내부 오류 - " + ex.getMessage(), ex);
```
로그에 스택트레이스 전에 message도 보고 싶어서 추가했습니다! 클라이언트에는 서버 내부 오류라는 것만 전달하면 되는데 로깅에서는 정확히 어떤 예외로부터 시작됐는지 알고싶을 것 같아서요

ex. S3예외면 서버 내부 오류 - S3 파일 업로드 실패: stackTrace출력어쩌구

이렇게 나오게 하고 싶었어요

yuyoungrhee commented on backend/src/main/resources/templates/admin/category-list.html:
```diff
@@ -8,7 +8,7 @@
 </head>
 <body>
-<div th:replace="admin/navbar :: navbar"></div>
+<div th:insert="~{admin/navbar :: navbar}"></div>
```
replace가 타임리프의 구버전 문법이라 변경했습니다. WARN 로그가 뜨더라구요

yuyoungrhee committed:
[BE] refactor: created status 반환 시 uri 제거 #208

- 안드로이드에서 uri 정보를 사용하지 않음

yuyoungrhee committed:
[BE] refactor: HearitDetailResponse에서 keyword의 id도 주도록 추가 #208

gabean13 commented on backend/src/main/java/com/onair/hearit/admin/application/AdminHearitService.java:
```diff
@@ -66,13 +69,36 @@ private Map<Long, List<KeywordInHearit>> mapKeywordsByHearitId(List<HearitKeywor
         saveHearitKeywords(request.keywordIds(), savedHearit);
     }
 
+    private void validateFiles(MultipartFile originalAudio, MultipartFile shortAudio, MultipartFile scriptFile) {
+        String originalFilename = originalAudio.getOriginalFilename();
+        String shortFileName = shortAudio.getOriginalFilename();
+        String scriptFileName = scriptFile.getOriginalFilename();
+
+        if (!originalFilename.startsWith("ORG")) {
+            throw new InvalidInputException("원본 오디오 파일명은 'ORG'로 시작해야 합니다.");
+        }
+
+        if (!shortFileName.startsWith("SHR")) {
+            throw new InvalidInputException("요약 오디오 파일명은 'SHR'로 시작해야 합니다.");
+        }
+
+        if (!scriptFileName.startsWith("SCR")) {
+            throw new InvalidInputException("스크립트 파일명은 'SCR'로 시작해야 합니다.");
+        }
+    }
```
혹시 동일 uuid가 아니면 막아주는 것도 검증하는 것 어떻게 생각하시나요? 그리고 확장자도 검증하면 좋을 거 같아요 이것도 걍 enum으로 처리하면 편리할 것 같긴해용

yuyoungrhee replied:
아맞다 같은 uuid검증빼먹었네요 확장자는 form에서막아주고있어요 그래도 서버에서 확인할게요
[[BE] refactor: 파일 업로드 및 검증 로직 리팩터링](https://github.com/woowacourse-teams/2025-hEARit/pull/216/commits/190a39f5761454550af4fbb0e2fe48423cd0d31f) https://github.com/woowacourse-teams/2025-hEARit/issues/208

gabean13 commented on backend/src/main/java/com/onair/hearit/admin/application/S3Uploader.java:
```diff
@@ -0,0 +1,56 @@
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class S3Uploader {
```
혹시 S3Uploader말고 ContentStorageService 같은 이름 어떤가요? S3인걸 클래스명에서 드러낼필요가 있나 해서요! 사용하는 측에서도 굳이 몰라도 되지 않나욤?

yuyoungrhee replied:
그러네요 인정합니다 변경할게욧~ FileStorageService는 어떠신지
[[BE] refactor: S3Uploader → FileStorageService로 클래스명 변경](https://github.com/woowacourse-teams/2025-hEARit/pull/216/commits/bd3b238bc435a9daff321929e854603730a3a5b4) https://github.com/woowacourse-teams/2025-hEARit/issues/208

gabean13 replied:
조아욧

gabean13 commented on backend/src/main/java/com/onair/hearit/admin/application/S3Uploader.java:
```diff
@@ -0,0 +1,56 @@
+    @Value("${aws.s3.bucket}")
+    private String bucket;
+
+    public String uploadOriginalAudio(MultipartFile multipartFile) {
+        String key = "hearit/audio/original/" + multipartFile.getOriginalFilename();
+        return uploadToS3(multipartFile, key);
+    }
+
+    public String uploadShortAudio(MultipartFile multipartFile) {
+        String key = "hearit/audio/short/" + multipartFile.getOriginalFilename();
+        return uploadToS3(multipartFile, key);
+    }
+
+    public String uploadScriptFile(MultipartFile multipartFile) {
+        String key = "hearit/script/" + multipartFile.getOriginalFilename();
+        return uploadToS3(multipartFile, key);
+    }
```
이 부분 관련있는 문자열들이라서 enum으로 관리하는 거 어떤가용?
SHT, SCR, ORG 랑 file 위치, 확장자 enum으로 관리하고 key를 계산하도록 하면 메서드 하나로도 해결할 수 있을 것 같아요

yuyoungrhee replied:
진짜똑똑이다 반영할게요
[[BE] refactor: 파일 업로드 및 검증 로직 리팩터링](https://github.com/woowacourse-teams/2025-hEARit/pull/216/commits/190a39f5761454550af4fbb0e2fe48423cd0d31f) https://github.com/woowacourse-teams/2025-hEARit/issues/208

gabean13 reviewed (commented):
리뷰 남겼어용~

yuyoungrhee committed:
[BE] refactor: 파일 업로드 및 검증 로직 리팩터링 #208

- 파일 유형별 prefix, 확장자, 업로드 경로를 FileType enum으로 통합 관리
- S3Uploader에서 파일 업로드 메서드 단일화 (uploadFile)
- AdminHearitService 내 파일 검증 로직을 FileValidator로 위임

yuyoungrhee committed:
[BE] refactor: S3Uploader → FileStorageService로 클래스명 변경 #208

- 파일 저장 책임을 보다 추상적인 관점에서 표현하기 위함

yuyoungrhee committed:
[BE] refactor: 관리 ㅏ히어릿 업로드 파일 null체크 추가 #208

yuyoungrhee committed:
[BE] refactor: 롬복 적용 등 컨벤션 반영 #208

gabean13 reviewed (dismissed):
고생하셨어용

Byesol reviewed (approved):
확인했습니다!

gabean13 reviewed (approved):
고생하셧어요
